### PR TITLE
Use API-lite for custom ops

### PR DESCRIPTION
**Pull request by souptc**
Add a header for the lite custom op api, this is built atop ortw::cusotmop api in extension folder and can works with legacy version of onnxruntime.

### CODE REVIEW

**Comment by souptc** [comment]
update the PR to include cv2 / image kernels.

Now we have two limitation remaining: 1) string tensor support. 2) support initialization.

**Comment by souptc** [comment]
Update the PR to support register a custom op kernerl struct (with construct and compute). now the only thing remain is the string tensor.

File: includes/custom_op_lite.h
Comment (wenbingl) [inline includes/custom_op_lite.h:1]: Is this the new API header, why is it lite?

File: includes/ocos.h [hunk 1]
```diff
@@ -85,6 +85,53 @@ class CuopContainer {
   std::vector<std::shared_ptr<OrtCustomOp>> op_instances_;  // use shared_ptr to capture type specific deleter
 };
 
+#define LiteCustomOp(name, f) []() { return std::shared_ptr<OrtCustomOp>(ortc::CreateCustomOpT2(name, f)); }
```
Comment (wenbingl) [inline includes/ocos.h:88]: can we implement these macro by some template function/class?

File: includes/onnxruntime_customop.hpp [hunk 2]
```diff
@@ -295,3 +295,7 @@ inline OrtValue* CustomOpApi::KernelContext_GetOutput(OrtKernelContext* context,
 }
 
 }  // namespace OrtW
+
+// !! TODO: only do it for legecy ort build
```
Comment (wenbingl) [inline includes/onnxruntime_customop.hpp:299]: I guess the logic should be implemented into CMakeLists.txt?

File: operators/math/math.cc [hunk 3]
```diff
@@ -6,12 +6,14 @@
 #include "segment_extraction.hpp"
 #include "segment_sum.hpp"
 
-
-FxLoadCustomOpFactory LoadCustomOpClasses_Math = 
-    LoadCustomOpClasses<CustomOpClassBegin, 
-                        CustomOpNegPos,
+const std::vector<const OrtCustomOp*>& MathLoader() {
+  static OrtOpLoader op_loader(LiteCustomOp("NegPos", neg_pos),
```
Comment (wenbingl) [inline operators/math/math.cc:10]: can all members which are stored in OrtOpLoader be released without any ABI from ORT?

File: operators/cv2/imgcodecs/imdecode.hpp [hunk 4]
```diff
@@ -12,70 +12,28 @@
...
-    // Setup output & copy to destination
-    const cv::Size decoded_image_size = decoded_image.size();
-    const int64_t colors = 3;
-
-    const std::vector<int64_t> output_dimensions{decoded_image_size.height, decoded_image_size.width, colors};
-    OrtValue* const output_value = ort_.KernelContext_GetOutput(
-        context, 0, output_dimensions.data(), output_dimensions.size());
-    uint8_t* const decoded_image_data = ort_.GetTensorMutableData<uint8_t>(output_value);
-    memcpy(decoded_image_data, decoded_image.data, decoded_image.total() * decoded_image.elemSize());
+void image_decoder(const ortc::TensorT<uint8_t>& input,
```
Comment (wenbingl) [inline operators/cv2/imgcodecs/imdecode.hpp:15]: How do you handle the optional input/output in new API?

Comment (souptc) [reply operators/cv2/imgcodecs/imdecode.hpp:15]: not supported yet.

File: includes/custom_op_lite.h
Comment (souptc) [reply includes/custom_op_lite.h:1]: it is not designed to replace the existing API, it is just a lite weight approach to simplify the kernel authoring for 80% cases.

File: includes/onnxruntime_customop.hpp (hunk 2 above)
Comment (souptc) [reply includes/onnxruntime_customop.hpp:299]: yes let's fix it later once the API is finalized.

**Comment by souptc** [comment]
migrate all the operators

**Review by RandySheriffH — COMMENTED** [review]
note - when ext upgrade supported ort version, it could use the lite custom op header from ort directly.

File: includes/custom_op_lite.h [hunk 5]
```diff
@@ -0,0 +1,668 @@
...
+                                          kernel->api_->KernelContext_GetOutputCount(context),
+                                          kernel->ep_);
+      std::apply([kernel](Args const&... t_args) { kernel->compute_fn_(t_args...); }, t);
+    };
+
+    OrtCustomOp::CreateKernel = [](const OrtCustomOp* this_, const OrtApi* ort_api, const OrtKernelInfo* info) {
+      auto kernel = std::make_unique<Kernel>();
+      auto self = static_cast<const OrtLiteCustomFunc*>(this_);
+      kernel->compute_fn_ = self->compute_fn_;
+      kernel->num_input_ = self->input_types_.size();
```
Comment (wenbingl) [inline includes/custom_op_lite.h:584]: May I have missed, but where num_input_ and num_output_ were used?

File: operators/text/re2_strings/string_regex_replace.cc [hunk 6]
```diff
@@ -70,15 +71,43 @@ const char* CustomOpStringRegexReplace::GetName() const { return "StringRegexRep
 size_t CustomOpStringRegexReplace::GetInputTypeCount() const {
   return 3;
 };
+*/
 
-ONNXTensorElementDataType CustomOpStringRegexReplace::GetInputType(size_t /*index*/) const {
-  return ONNX_TENSOR_ELEMENT_DATA_TYPE_STRING;
-};
+//ONNXTensorElementDataType CustomOpStringRegexReplace::GetInputType(size_t /*index*/) const {
```
Comment (wenbingl) [inline operators/text/re2_strings/string_regex_replace.cc:76]: should this code block be removed?

File: operators/text/string_ecmaregex_split.cc [hunk 7]
```diff
@@ -15,40 +15,26 @@ KernelStringECMARegexSplitWithOffsets::KernelStringECMARegexSplitWithOffsets(con
   ignore_case_ = TryToGetAttributeWithDefault("ignore_case", false);
 }
 
-void KernelStringECMARegexSplitWithOffsets::Compute(OrtKernelContext* context) {
+void KernelStringECMARegexSplitWithOffsets::Compute(const ortc::Tensor<std::string>& input,
+                                                    std::string_view pattern,
```
Comment (wenbingl) [inline operators/text/string_ecmaregex_split.cc:19]: can these inputs be "const string_view&" ?

Comment (RandySheriffH) [reply operators/text/string_ecmaregex_split.cc:19]: string_view is read-only already.

File: operators/text/re2_strings/string_regex_replace.cc (hunk 6 above)
Comment (RandySheriffH) [reply operators/text/re2_strings/string_regex_replace.cc]: Removed.

File: includes/custom_op_lite.h (hunk 5 above)
Comment (RandySheriffH) [reply includes/custom_op_lite.h]: They are obsolete, and removed.

File: operators/tokenizer/clip_tokenizer.cc [hunk 8]
```diff
@@ -107,21 +108,22 @@ std::vector<int64_t> KernelClipBpeTokenizer::Tokenize(ustring& input, int64_t ma
     // Add offset mapping for EOS token
     offset_mapping.emplace_back(std::make_pair(0, 0));
 
-    // Add offset mappings for input in this instance to list of offset mappings for all inputs
+    // Add offset mappings for input in this instance to list of offset mappings for all inputsclip_tokenizer.cc
```
[the remark this answers is not in the document]
Comment (RandySheriffH) [reply operators/tokenizer/clip_tokenizer.cc]: Corrected.

File: operators/math/dlib/stft_norm.hpp [hunk 9]
```diff
@@ -6,38 +6,31 @@
 #include "ocos.h"
 #include <dlib/matrix.h>
 
-struct KernelStft : BaseKernel {
-  KernelStft(const OrtApi& api, const OrtKernelInfo& info, bool return_magnitude)
-      : BaseKernel(api, info), return_magnitude_(return_magnitude) {
+template <bool with_norm = false>
```
[the remark this answers is not in the document]
Comment (RandySheriffH) [reply operators/math/dlib/stft_norm.hpp]: Switched to constructor arg instead.

File: includes/custom_op_lite.h [hunk 10]
```diff
@@ -0,0 +1,668 @@
...
+      return std::tuple_cat(current, next);                                                                                                                           \
+    } else {                                                                                                                                                          \
+      std::tuple<T> current = std::tuple<T>{};                                                                                                                        \
+      auto next = CreateTuple<ith_input, ith_output + 1, Ts...>(api, context, tensors, num_input, num_output, ep);                                                    \
+      return std::tuple_cat(current, next);                                                                                                                           \
+    }                                                                                                                                                                 \
+  }
+#define CREATE_TUPLE(data_type) \
```
[the remark this answers is not in the document]
Comment (RandySheriffH) [reply includes/custom_op_lite.h:419]: There are too many types to stack up here, code would be unreadable.
Even with the macro, one could still set break point and step in.

File: includes/custom_op_lite.h [hunk 11]
```diff
@@ -0,0 +1,668 @@
...
+ protected:
+  const OrtW::CustomOpApi& api_;
+  OrtKernelContext& ctx_;
+  size_t indice_;
+  bool is_input_;
+  std::optional<std::vector<int64_t>> shape_;
+};
+
+template <typename T>
+struct Span {
```
[the remark this answers is not in the document]
Comment (RandySheriffH) [reply includes/custom_op_lite.h:63]: Moved one of is member "Data" to "data" to align with std::span.

File: includes/custom_op_lite.h [hunk 12]
```diff
@@ -0,0 +1,668 @@
+#pragma once
+#include "onnxruntime_customop.hpp"
+#include <optional>
+#include <numeric>
+#define SUPPORT_ORT_API_VERSION_TO 13
```
[the remark this answers is not in the document]
Comment (RandySheriffH) [reply includes/custom_op_lite.h:9]: Added:
"uplevel the version when supported ort version migrates to newer ones"

File: includes/custom_op_lite.h [hunk 13]
```diff
@@ -0,0 +1,668 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (RandySheriffH) [reply includes/custom_op_lite.h:4]: Done.

File: test/test_tools_add_pre_post_processing_to_model.py [hunk 14]
```diff
@@ -495,10 +495,7 @@ def create_pipeline_and_run_for_nms(self, output_model: Path, length: int,
         create_named_value = pre_post_processing.utils.create_named_value
 
         inputs = [create_named_value("box_and_score", onnx.TensorProto.FLOAT, ["num_boxes", length])]
-
-        onnx_opset = 16
```
[the remark this answers is not in the document]
Comment (RandySheriffH) [reply test/test_tools_add_pre_post_processing_to_model.py:499]: Removed main, should  be fine. now.

[the remark this answers is not in the document]
Comment (wenbingl) [reply test/test_tools_add_pre_post_processing_to_model.py:499]: I still see the changes here, are they expected?

File: operators/vision/draw_bounding_box.cc [hunk 15]
```diff
@@ -205,7 +205,7 @@ void DrawBoxesForNumClasses(ImageView& image, const BoxArray& boxes, int64_t thi
             [](const std::pair<size_t, int64_t>& first_, const std::pair<size_t, int64_t>& second_) {
               return first_.second < second_.second;
             });
-  for (int64_t i = static_cast<int64_t>(box_reverse.size()) - 1; i >= 0; --i) {
```
Comment (wenbingl) [inline operators/vision/draw_bounding_box.cc:208]: merge error?

**Review by wenbingl — COMMENTED** [review]
Somehow, I still can see some changes in this PR are reverting other's, please fix that before it.

File: operators/vision/draw_bounding_box.cc (hunk 15 above)
Comment (RandySheriffH) [reply operators/vision/draw_bounding_box.cc:208]: Main remerged, the line is there now.

File: test/test_tools_add_pre_post_processing_to_model.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (RandySheriffH) [reply test/test_tools_add_pre_post_processing_to_model.py:499]: Main remerged, the line restored.

**Review by wenbingl — APPROVED** [review]

**Pull request merged**